### PR TITLE
Issue #413: `advance` suggests a push to remote

### LINKED ISSUE #413: Shouldn't `advance` suggest a push to remote as well?
I literally don't remember doing `git machete advance` that would NOT be followed by `git push` on the fast-forwarded branch (which is btw almost always `develop` or `master`).

Currently `advance` fast-forwards current branch to a child and subsequently asks whether the child should be slid out.
I'd say the push should be suggested in-between: 
1. fast-forward current branch to a child
2. ask if push current branch
3. ask if slide out the child (ask even if push has been rejected; **don't ask** if push has been accepted and **failed**).

The reason for putting push before slide-out is a very common situation when it turns out that the PR for branch doesn't satisfy Github's conditions/branch protection rules (i.e. some new commits just appeared on `develop`). So a push is declined... in such a situation, still having the child branch around is pretty useful. As a killer feature, undoing of the fast-forward could be suggested instead!

### PR DESCRIPTION
**Pull request by amalota**
closes #413

### CODE REVIEW

**Comment by PawelLipski** [comment]
Actually... pls move it to v3.9.0 milestone! to avoid bloating v3.8.0 (I'd like to make a release soon after `github sync` is ready)

File: docs/source/cli_help/advance.rst [hunk 1]
```diff
@@ -14,7 +14,7 @@ advance
 
     git machete advance [-y|--yes]
 
-Fast forwards (as in ``git merge --ff-only``) the current branch ``C`` to match its downstream ``D``,
+Fast forwards (as in ``git merge --ff-only``) the current branch ``C`` to match its downstream ``D``, pushes current branch
```
Comment (PawelLipski) [inline docs/source/cli_help/advance.rst]:
```suggestion
Fast forwards (as in ``git merge --ff-only``) the current branch ``C`` to match its downstream ``D``, pushes ``C``
```

File: docs/source/cli_help/advance.rst [hunk 2]
```diff
@@ -14,7 +14,7 @@ advance
 
     git machete advance [-y|--yes]
 
-Fast forwards (as in ``git merge --ff-only``) the current branch ``C`` to match its downstream ``D``,
```
Comment (PawelLipski) [inline docs/source/cli_help/advance.rst:17]: Pls update docs.py as well

File: RELEASE_NOTES.md [hunk 3]
```diff
@@ -4,6 +4,7 @@
 ## New in git-machete 3.8.0
 
 - added: `--all`, `--mine`, `--by` flags and parameter `<PR-number-1> ... <PR-number-N>` to `git machete github checkout-prs`
+- added: `advance` command now also pushes the branch after the merge
```
Comment (PawelLipski) [inline RELEASE_NOTES.md]: v3.9.0 FTW

File: git_machete/client.py [hunk 4]
```diff
@@ -602,8 +602,14 @@ def connected_with_green_edge(bd: LocalBranchShortName) -> bool:
...
 
-        ans = self.ask_if(f"\nBranch {bold(down_branch)} is now merged into {bold(branch)}. Slide {bold(down_branch)} out of the tree of branch dependencies?" + get_pretty_choices('y', 'N'),
-                          f"\nBranch {bold(down_branch)} is now merged into {bold(branch)}. Sliding {bold(down_branch)} out of the tree of branch dependencies...", opt_yes=opt_yes)
+        remote = self.__git.get_strict_remote_for_fetching_of_branch(branch)
+        ans = self.ask_if(f"\nBranch {bold(branch)} is now fast-forwarded to match {bold(down_branch)}. Push {bold(branch)} to {bold(remote)}?" + get_pretty_choices('y', 'N'),
+                          f"\nBranch {bold(branch)} is now fast-forwarded to match {bold(down_branch)}. Pushing {bold(branch)} to {bold(remote)}...", opt_yes=opt_yes)
+        if ans in ('y', 'yes'):
+            self.__git.push(remote, branch)
+
+        ans = self.ask_if(f"\nBranch {bold(branch)} is now pushed to {bold(remote)}. Slide {bold(down_branch)} out of the tree of branch dependencies?" + get_pretty_choices('y', 'N'),
```
Comment (PawelLipski) [inline git_machete/client.py]: Hmmm this doesn't need to be true, what if the user chose `N` to `Push...`?

Comment (PawelLipski) [reply git_machete/client.py]: I'd leave it as it was, with `...is now merged...`

Comment (amalota) [reply git_machete/client.py]: @PawelLipski I added new logic to change the msg depending on the answer.

File: docs/source/cli_help/advance.rst [hunk 5]
```diff
@@ -14,7 +14,7 @@ advance
...
+Fast forwards (as in ``git merge --ff-only``) the current branch ``C`` to match its downstream ``D``, pushes ``C``
 and subsequently slides out ``D``. Both steps require manual confirmation unless ``-y/--yes`` is provided.
```
Comment (PawelLipski) [inline docs/source/cli_help/advance.rst]:
```suggestion
and subsequently slides out ``D``. All three steps require manual confirmation unless ``-y/--yes`` is provided.
```

File: docs/source/cli_help/advance.rst [hunk 6]
```diff
@@ -14,7 +14,7 @@ advance
 
     git machete advance [-y|--yes]
 
-Fast forwards (as in ``git merge --ff-only``) the current branch ``C`` to match its downstream ``D``,
+Fast forwards (as in ``git merge --ff-only``) the current branch ``C`` to match its downstream ``D``, pushes ``C``
```
Comment (PawelLipski) [inline docs/source/cli_help/advance.rst:17]: Pls reflect the changes in docs.py

File: docs/source/cli_help/advance.rst [hunk 7]
```diff
@@ -14,7 +14,7 @@ advance
...
 and subsequently slides out ``D``. Both steps require manual confirmation unless ``-y/--yes`` is provided.
 
 The downstream ``C`` is selected according to the following criteria:
```
Comment (PawelLipski) [inline docs/source/cli_help/advance.rst]: My old mistake:
```suggestion
The downstream ``D`` is selected according to the following criteria:
```

Comment (PawelLipski) [reply docs/source/cli_help/advance.rst]: Also, `If --yes is specified...` is a part of the third point ` if C has more than one downstream...`, pls move

File: git_machete/client.py [hunk 8]
```diff
@@ -602,8 +602,17 @@ def connected_with_green_edge(bd: LocalBranchShortName) -> bool:
...
                 return
 
-        ans = self.ask_if(f"\nBranch {bold(down_branch)} is now merged into {bold(branch)}. Slide {bold(down_branch)} out of the tree of branch dependencies?" + get_pretty_choices('y', 'N'),
-                          f"\nBranch {bold(down_branch)} is now merged into {bold(branch)}. Sliding {bold(down_branch)} out of the tree of branch dependencies...", opt_yes=opt_yes)
+        remote = self.__git.get_strict_remote_for_fetching_of_branch(branch)
+        ans = self.ask_if(f"\nBranch {bold(branch)} is now fast-forwarded to match {bold(down_branch)}. Push {bold(branch)} to {bold(remote)}?" + get_pretty_choices('y', 'N'),
+                          f"\nBranch {bold(branch)} is now fast-forwarded to match {bold(down_branch)}. Pushing {bold(branch)} to {bold(remote)}...", opt_yes=opt_yes)
+        if ans in ('y', 'yes'):
+            self.__git.push(remote, branch)
+            branch_pushed_or_fast_forwarded_msg = f"\nBranch {bold(branch)} is now pushed to {bold(remote)}."
```
Comment (PawelLipski) [inline git_machete/client.py:614]:
```suggestion
            branch_pushed_or_fast_forwarded_msg = f"\nBranch {bold(branch)} is now fast-forwarded and pushed to {bold(remote)}."
```

Coz suggesting slide-out of `down_branch` only makes sense once `branch` is fast-forwarded to `down_branch`! Pushing `branch` is incidental/unrelated to sliding-out of `down_branch`.

Comment (PawelLipski) [reply git_machete/client.py:614]: So we'd better mention the fast-forward as the reason for slide-out in both cases (push or no push)

Comment (PawelLipski) [reply git_machete/client.py:614]: Awww okay I see now:

image

So it's good! let's leave the current version

**Review by PawelLipski — APPROVED** [review]
LGTM! let's refrain from merging as for now, let's release 3.8.0 first

**Comment by PawelLipski** [comment]
Lemme rebase and merge it as it's long ready, just waited for v3.8.0 to be released...

**Pull request merged**